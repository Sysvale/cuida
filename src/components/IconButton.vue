<template>
	<span id="cds-icon-button">
		<component
			:is="tooltipText ? 'cds-tooltip' : 'span'"
			:text="innerTooltipText"
		>
			<button
				class="cds-icon-button__container"
				:class="computedModifiers"
				@click="clickHandler"
			>
				<cds-icon
					:key="inputControlPanel"
					:name="internalIcon"
					class="cds-icon-button__icon"
				/>
			</button>
		</component>
	</span>
</template>

<script>
import CdsIcon from './Icon.vue';
import CdsTooltip from './Tooltip.vue';

const predefinedSizes = [
	'sm',
	'md',
	'lg',
];

export default {
	components: {
		CdsIcon,
		CdsTooltip,
	},

	props: {
		/**
		 * Especifica o tamanho do botão. São 3 tamanhos implementados: 'sm', 'md', 'lg'.
		 */
		size: {
			type: String,
			default: 'md',
			validator: (value) => {
				return predefinedSizes.includes(value);
			},
		},
		/**
		* Especifica o `name` do ícone do cuida icons a ser apresentado no corpo do botão.
		*/
		icon: {
			type: String,
			default: 'create-outline',
		},
		/**
		 * Controla a disponibilidade do botão.
		 */
		disabled: {
			type: Boolean,
			default: false,
		},
		/**
		* Texto a ser exibido como tooltip com o hover do botão.
		*/
		tooltipText: {
			type: String,
			default: null,
		},
		/**
		* Quando ativo, faz com que, após iteração do usuário, o ícone do botão seja temporariamente alterado.
		*/
		feedbackOnClick: {
			type: Boolean,
			default: false,
		},
		/**
		* Ícone mostrado após clique do usuário.
		*/
		feedbackIcon: {
			type: String,
			default: 'check-outline',
		},
		/**
		* A variante de cor. São 9 variantes implementadas: 'green', 'teal',
		* 'blue', 'indigo', 'violet', 'pink', 'red', 'orange', 'amber', 'gray' e 'dark'.
		*/
		variant: {
			type: String,
			default: 'white',
		},
	},

	data() {
		return {
			predefinedSizes,
			innerTooltipText: this.tooltipText,
			internalIcon: this.icon,
			inputControlPanel: 0,
		};
	},

	computed: {
		predefinedSize() {
			return `cds-icon-button--${this.size}`;
		},

		computedModifiers() {
			const status = this.disabled ? 'cds-icon-button--disabled' : '';
			const variantClass = `cds-icon-button__container--${this.variant}`;

			return `${status} ${this.predefinedSize} ${variantClass}`;
		},
	},

	watch: {
		tooltipText() {
			this.innerTooltipText = this.tooltipText;
		},
	},

	methods: {
		clickHandler(e) {
			if (this.disabled) {
				return;
			}

			if (this.feedbackOnClick) {
				this.inputControlPanel += 1;
				this.internalIcon = this.feedbackIcon;
				setTimeout(() => {
					this.internalIcon = this.icon;
					this.inputControlPanel += 1;
				}, 1000);

				this.inputControlPanel += 1;
			}

			/**
			* Evento que indica que o botão foi clicado
			* @event cds-click
			* @type {Event}
			*/
			this.$emit('cds-click', e);
		},
	},
};
</script>

<style lang="scss" scoped>
@import '../assets/sass/tokens.scss';

.cds-icon-button {
	&__container {
		border: none !important;
		border-radius: $border-radius-extra-small;
		cursor: pointer;
		box-sizing: border-box;	

		@include variantResolver using ($color-name, $shade-50, $shade-100, $shade-200, $shade-300, $base-color, $shade-500, $shade-600) {
			@extend .cds-icon-button__container;
			background-color: $base-color;
			color: $n-0;

			@if ($color-name == 'gray') {
				background-color: $shade-100;
				color: $n-600;
			}

			@if ($color-name == 'white') {
				color: $n-600;
				outline: 1px solid $n-50 !important;

				&:hover {
					background-color: $n-10;
				}
			} @else {
				&:hover {
					background-color: $shade-500;
				}

				&:focus {
					outline: none !important;
				}
			}
		}
	}

	&--disabled {
		box-sizing: border-box;
		outline: 1px solid transparent !important;
		color: $n-300 !important;
		background-color: $n-20 !important;
		cursor: default !important;

		&:hover {
			outline: 1px solid transparent !important;
		}
	}

	&__icon {
		display: flex;
	}
}

.cds-icon-button {
	&--sm {
		padding: pYX(2, 2);
		border-radius: $border-radius-lil;

		.cds-icon-button__icon {
			transition: $hover;
			width: 16px;
			height: 16px;
		}
	}

	&--md {
		padding: pYX(2, 2);
		border-radius: $border-radius-extra-small;

		.cds-icon-button__icon {
			transition: $hover;
			width: 20px;
			height: 20px;
		}
	}

	&--lg {
		padding: pYX(2, 2);
		border-radius: $border-radius-extra-small;

		.cds-icon-button__icon {
			transition: $hover;
			width: 24px;
			height: 24px;
		}
	}
}
</style>
