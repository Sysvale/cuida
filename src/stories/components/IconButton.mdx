import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import IconButton from '../../components/IconButton.vue';
import * as IconButtonStories from './IconButton.stories';

<Meta of={IconButtonStories} />

export const Template = (args) => ({
	components: { CdsIconButton: IconButton },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-icon-button
			v-bind="args"
			@cds-click="logIconButtonClick"
		/>`,
	methods: {
		logIconButtonClick(event) {
			console.info('%con click: ', 'color: #2C70CD;', event);
		},
	},
});

# IconButton

### IconButtons são componentes que permitem que o usuário execute uma ação com um toque.

***

<br />

## Quando usar:

* For necessário comunicar ao usuário que ele pode executar uma ação na interface,
  seja em diálogos, janelas modais, formulários, cards, etc.

<br />

## Quando não usar:

* For necessário dar a opção de executar uma ação como um detalhe pequeno na interface.
  Nesse caso, recomendamos o uso de links.

<br />

## Preview

<Canvas>
  <Story of={IconButtonStories.IconButton} />
</Canvas>

<Controls />
