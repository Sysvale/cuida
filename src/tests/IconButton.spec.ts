import { describe, test, expect } from 'vitest';
import IconButton from '../components/IconButton.vue';
import { mount } from '@vue/test-utils';

describe('IconButton', () => {
	test('renders correctly', async () => {
		const wrapper = mount(IconButton, {
			global: {
				stubs: {
					'cds-icon': true,
					'cds-tooltip': true,
				},
			},
			props: {
				size: 'lg',
				icon: 'trash-outline'
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
