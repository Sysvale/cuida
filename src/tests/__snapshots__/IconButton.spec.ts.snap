// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IconButton > renders correctly 1`] = `"<span data-v-f6045c74="" id="cds-icon-button"><span data-v-f6045c74=""><button data-v-f6045c74="" class="cds-icon-button__container cds-icon-button--lg cds-icon-button__container--white"><cds-icon-stub data-v-f6045c74="" name="trash-outline" width="24" height="24" color="currentColor" dark="false" light="false" class="cds-icon-button__icon"></cds-icon-stub></button></span></span>"`;
